batchproducer/batchproducer.go

```diff
@@ -223,15 +235,44 @@ func (b *batchProducer) run() {
 	}
 }
 
-// from/for interface BatchProducer
+// from/for interface Producer
 func (b *batchProducer) Stop() error {
-	// TODO: Immediately stop accepting new records by Add, then block until all the records in the buffer have been sent
 	if b.isRunning() {
 		b.stop <- true
 	}
 	return nil
 }
 
+// from/for interface Producer
+// TODO: send all batches in parallel, will require broader refactoring
+func (b *batchProducer) Flush(timeout time.Duration) (int, error) {
+	b.Stop()
+
+	timer := time.NewTimer(timeout)
+	if timeout == 0 {
+		timer.Stop()
+	}
+
+	timedOut := false
+
+loop:
+	for len(b.records) > 0 {
+		select {
+		case <-timer.C:
+			timedOut = true
+			break loop
```

Comment:
I'm having a bit of trouble understanding this block of code.

While we have more records, we want to send the stats in batches. However, if we've timed out, we jump  back up to the beginning of the loop which will continue to send batches even after it has "timed out" since `timer.C` represents a single event.

If I'm understanding this correctly, wouldn't we want to check `timedOut` after the `select` block and then `break` out of the `for` loop?